pkg/storage/stores/shipper/indexshipper/indexgateway/gateway.go

```diff
@@ -340,6 +344,261 @@ func (g *Gateway) GetVolume(ctx context.Context, req *logproto.VolumeRequest) (*
 	return g.indexQuerier.Volume(ctx, instanceID, req.From, req.Through, req.GetLimit(), req.TargetLabels, req.AggregateBy, matchers...)
 }
 
+func (g *Gateway) GetShards(request *logproto.ShardsRequest, server logproto.IndexGateway_GetShardsServer) error {
+	ctx := server.Context()
+	log, _ := spanlogger.New(context.Background(), "IndexGateway.GetShards")
+	defer log.Finish()
+
+	instanceID, err := tenant.TenantID(ctx)
+	if err != nil {
+		return err
+	}
+
+	p, err := ExtractShardRequestMatchersAndAST(request.Query)
```

Comment:
As my previous comment, we can avoid this by having the Plan inside the request.